File: MalomAPI_cpp/PerfectPlayer.cpp
// Malom, a Nine Men's Morris (and variants) player and solver program.
// Copyright(C) 2007-2016  Gabor E. Gevay, Gabor Danner
// Copyright (C) 2023 The Sanmill developers (see AUTHORS file)
//
// See our webpage (and the paper linked from there):
// http://compalg.inf.elte.hu/~ggevay/mills/index.php
//
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

#include <fstream>
#include <iostream>
#include <map>
#include <stdexcept>
#include <string>
#include <bitset>
#include <vector>
#include <cassert> // for assert
#include <cstdint> // for int64_t
#include <cstdlib> // for std::exit
#include <exception> // for std::exception
#include <iostream> // for std::cerr
#include <mutex> // for std::mutex and std::lock_guard
#include <stdexcept> // for std::out_of_range

class id {
    // Define your class 'id' here
};

class Sector {
public:
    Sector(id _id)
    {
        // Define your constructor here
    }
};

class Sectors {
public:
    static std::map<id, Sector> sectors;
    static bool created;

    static std::map<id, Sector> getsectors()
    {
        try {
            if (!created) {
                Wrappers::Init::init_sym_lookuptables();
                Wrappers::Init::init_sec_vals();

                for (int w = 0; w <= Rules::MaxKSZ; ++w) {
                    for (int b = 0; b <= Rules::MaxKSZ; ++b) {
                        for (int wf = 0; wf <= Rules::MaxKSZ; ++wf) {
                            for (int bf = 0; bf <= Rules::MaxKSZ; ++bf) {
                                std::string fname = Rules::VariantName + "_" + std::to_string(w) + "_" + std::to_string(b) + "_" + std::to_string(wf) + "_" + std::to_string(bf) + ".sec" + Constants::Fname_suffix;
                                // std::cout << "Looking for database file " << fname << std::endl;
                                id _id(w, b, wf, bf);
                                std::ifstream file(fname);
                                if (file.good()) {
                                    sectors[_id] = Sector(_id);
                                }
                            }
                        }
                    }
                }
                created = true;
            }
            return sectors;
        } catch (std::exception& ex) {
            if (dynamic_cast<std::out_of_range*>(&ex)) {
                throw;
            }
            std::cerr << "An error happened in getsectors\n"
                      << ex.what() << std::endl;
            exit(1);
        }
    }

    static bool HasDatabase()
    {
        return getsectors().size() > 0;
    }
};

// Initialize static member variables
std::map<id, Sector> Sectors::sectors;
bool Sectors::created = false;

class PerfectPlayer : public Player {
public:
    std::map<id, Sector> secs;

    PerfectPlayer()
    {
        assert(Sectors::HasDatabase());
        secs = Sectors::getsectors();
    }

    void Enter(Game _g) override
    {
        Player::Enter(_g);
    }

    void Quit() override
    {
        Player::Quit();
    }

    Sector* PerfectPlayer::GetSec(GameState s)
    {
        try {
            if (s.KLE)
                return nullptr;

            id id_val(s.StoneCount[0], s.StoneCount[1], Rules::MaxKSZ - s.SetStoneCount[0], Rules::MaxKSZ - s.SetStoneCount[1]);

            if (s.SideToMove == 1) {
                id_val.negate();
            }

            return &secs[id_val];
        } catch (std::exception& ex) {
            if (typeid(ex) == typeid(std::out_of_range))
                throw;
            std::cerr << "An error happened in GetSec\n"
                      << ex.what() << std::endl;
            std::exit(1);
        }
        return nullptr;
    }

    std::string PerfectPlayer::ToHumanReadableEval(gui_eval_elem2 e)
    {
        try {
            return e.ToString();
        } catch (std::exception& ex) {
            std::cerr << "An error happened in ToHumanReadableEval\n"
                      << ex.what() << std::endl;
            std::exit(1);
        }
        return "";
    }

    enum class MoveType {
        SetMove,
        SlideMove // should be renamed to SlideOrJumpMove
    };

    struct Move {
        int hon, hov;
        MoveType movetype;
        bool withtaking, onlytaking; // withtaking includes the steps in mill closure, onlytaking only includes removal
        int takehon;

        int toBitBoard()
        {
            if (onlytaking) {
                return 1 << takehon;
            }
            int ret = 1 << hov;
            if (movetype == MoveType::SlideMove) {
                ret += 1 << hon;
            }
            if (withtaking) {
                ret += 1 << takehon;
            }
            return ret;
        }
    };

    int futureKorongCount(GameState& s)
    {
        return s.stoneCount(s.sideToMove) + Rules.maxKSZ - s.setStoneCount(s.sideToMove); // TODO: refactor to call to futureStoneCount
    }

    bool makesMill(GameState& s, int hon, int hov)
    {
        GameState s2 = s;
        if (hon != -1)
            s2.T(hon) = -1;
        s2.T(hov) = s.sideToMove;
        return -1 != Rules.malome(hov, s2);
    }

    bool isMill(GameState& s, int m)
    {
        return -1 != Rules.malome(m, s);
    }

    std::vector<Move> setMoves(GameState& s)
    {
        std::vector<Move> r;
        for (int i = 0; i < 24; ++i) {
            if (s.T(i) == -1) {
                r.push_back(Move { i, false, makesMill(s, -1, i), MoveType::SetMove });
            }
        }
        return r;
    }

    std::vector<Move> slideMoves(GameState& s)
    {
        std::vector<Move> r;
        for (int i = 0; i < 24; ++i) {
            for (int j = 0; j < 24; ++j) {
                if (s.T(i) == s.sideToMove && s.T(j) == -1 && (futureKorongCount(s) == 3 || Rules.boardGraph(i, j))) {
                    r.push_back(Move { i, j, makesMill(s, i, j), false, MoveType::SlideMove, 0 });
                }
            }
        }
        return r;
    }

    // m has a withtaking step, where takehon is not filled out. This function creates a list, the elements of which are copies of m supplemented with one possible removal each.
    std::vector<Move> withTakingMoves(GameState& s, Move& m)
    {
        std::vector<Move> r;
        bool everythingInMill = true;
        for (int i = 0; i < 24; ++i) {
            if (s.T(i) == 1 - s.sideToMove && !isMill(s, i)) {
                everythingInMill = false;
            }
        }

        for (int i = 0; i < 24; ++i) {
            if (s.T(i) == 1 - s.sideToMove && (!isMill(s, i) || everythingInMill)) {
                Move m2 = m;
                m2.takehon = i;
                r.push_back(m2);
            }
        }
        return r;
    }

    std::vector<Move> onlyTakingMoves(GameState& s)
    { // there's some copy-paste code here
        std::vector<Move> r;
        bool everythingInMill = true;
        for (int i = 0; i < 24; ++i) {
            if (s.T(i) == 1 - s.sideToMove && !isMill(s, i)) {
                everythingInMill = false;
            }
        }

        for (int i = 0; i < 24; ++i) {
            if (s.T(i) == 1 - s.sideToMove && (!isMill(s, i) || everythingInMill)) {
                r.push_back(Move { 0, 0, false, true, MoveType::SlideMove, i }); // Assuming default values for hon, hov, and movetype
            }
        }
        return r;
    }

    std::vector<Move> getMoveList(GameState& s)
    {
        std::vector<Move> ms0, ms;
        if (!s.KLE) {
            if (Wrappers.Constants.variant == Wrappers.Constants.Variants.std || Wrappers.Constants.variant == Wrappers.Constants.Variants.mora) {
                if (s.setStoneCount(s.sideToMove) < Rules.maxKSZ) {
                    ms0 = setMoves(s);
                } else {
                    ms0 = slideMoves(s);
                }
            } else { // Lasker
                ms0 = slideMoves(s);
                if (s.setStoneCount(s.sideToMove) < Rules.maxKS Z) {
                    std::vector<Move> setMovesResult = setMoves(s);
                    ms0.insert(ms0.end(), setMovesResult.begin(), setMovesResult.end());
                }
            }

            for (int i = 0; i < ms0.size(); ++i) {
                if (!ms0[i].withtaking) {
                    ms.push_back(ms0[i]);
                } else {
                    std::vector<Move> withTakingMovesResult = withTakingMoves(s, ms0[i]);
                    ms.insert(ms.end(), withTakingMovesResult.begin(), withTakingMovesResult.end());
                }
            }
        } else { // KLE
            ms = onlyTakingMoves(s);
        }
        return ms;
    }

    GameState makeMoveInState(GameState& s, Move& m)
    {
        GameState s2(s);
        if (!m.onlyTaking) {
            if (m.moveType == MoveType::SetMove) {
                s2.makeMove(SetKorong(m.hov)); // Assuming the definition of SetKorong class is available
            } else {
                s2.makeMove(MoveKorong(m.hon, m.hov)); // Assuming the definition of MoveKorong class is available
            }
            if (m.withTaking)
                s2.makeMove(LeveszKorong(m.takeHon)); // Assuming the definition of LeveszKorong class is available
        } else {
            s2.makeMove(LeveszKorong(m.takeHon));
        }
        return s2;
    }

    // Assuming gui_eval_elem2 and getSec functions are defined somewhere
    gui_eval_elem2 moveValue(GameState& s, Move& m)
    {
        try {
            return eval(makeMoveInState(s, m)).undo_negate(getSec(s));
        } catch (const std::exception& ex) {
            std::cerr << "Exception in MoveValue\n"
                      << ex.what() << std::endl;
            std::exit(1);
        }
    }

    template <typename T, typename K>
    std::vector<T> allMaxBy(std::function<K(T)> f, std::vector<T>& l, K minValue)
    {
        std::vector<T> r;
        K ma = minValue;
        for (auto& m : l) {
            K e = f(m);
            if (e > ma) {
                ma = e;
                r.clear();
                r.push_back(m);
            } else if (e == ma) {
                r.push_back(m);
            }
        }
        return r;
    }

    // Assuming the definition of gui_eval_elem2::min_value function
    std::vector<Move> goodMoves(GameState& s)
    {
        return allMaxBy([s](Move m) { return moveValue(s, m); }, getMoveList(s), gui_eval_elem2::min_value(getSec(s)));
    }

    int NGMAfterMove(GameState& s, Move& m)
    {
        return numGoodMoves(makeMoveInState(s, m)); // Assuming numGoodMoves function is defined
    }

    template <typename T>
    T chooseRandom(std::vector<T>& l)
    {
        std::random_device rd;
        std::mt19937 gen(rd());
        std::uniform_int_distribution<> dis(0, l.size() - 1);
        return l[dis(gen)];
    }

    void sendMoveToGUI(Move& m)
    {
        if (!m.onlyTaking) {
            if (m.moveType == MoveType::SetMove) {
                G.makeMove(SetKorong(m.hov)); // Assuming the definition of SetKorong class is available
            } else {
                G.makeMove(MoveKorong(m.hon, m.hov)); // Assuming the definition of MoveKorong class is available
            }
        } else {
            G.makeMove(LeveszKorong(m.takeHon)); // Assuming the definition of LeveszKorong class is available
        }
    }

    void toMove(GameState& s)
    {
        try {
            Move mh = chooseRandom(goodMoves(s));
            sendMoveToGUI(mh);
        } catch (const std::out_of_range&) {
            sendMoveToGUI(chooseRandom(getMoveList(s)));
        } catch (const std::exception& ex) {
            std::cerr << "Exception in ToMove\n"
                      << ex.what() << std::endl;
            std::exit(1);
        }
    }

    int numGoodMoves(GameState& s)
    {
        if (futureKorongCount(s) < 3)
            return 0; // Assuming futureKorongCount function is defined
        auto ma = gui_eval_elem2::min_value(getSec(s)); // Assuming getSec function is defined
        Move mh;
        int c = 0;
        for (auto& m : getMoveList(s)) {
            auto e = moveValue(s, m);
            if (e > ma) {
                ma = e;
                mh = m;
                c = 1;
            } else if (e == ma) {
                c++;
            }
        }
        return c;
    }

    int cp;

    struct MoveValuePair {
        Move m;
        double val;
    };

    const double WRGMInf = 2; // Is this good?

    std::mutex evalLock;

    gui_eval_elem2 eval(GameState& s)
    {
        try {
            std::lock_guard<std::mutex> lock(evalLock);
            assert(!s.KLE); // Assuming s has a boolean member KLE

            id Id(s.stoneCount[0], s.stoneCount[1], Rules::MaxKSZ - s.setStoneCount[0], Rules::MaxKSZ - s.setStoneCount[1]);

            if (futureKorongCount(s) < 3)
                return gui_eval_elem2::virt_loss_val;

            int64_t a = 0;
            for (int i = 0; i < 24; ++i) {
                if (s.T[i] == 0) {
                    a |= (1ll << i);
                } else if (s.T[i] == 1) {
                    a |= (1ll << (i + 24));
                }
            }

            if (s.sideToMove == 1) {
                a = boardNegate(a);
                id.negate();
            }

            sec sec = secs[id];

            return sec.hash(a).item2;
        } catch (const std::exception& ex) {
            if (typeid(ex) == typeid(std::out_of_range))
                throw;
            std::cerr << "Exception in Eval\n"
                      << ex.what() << std::endl;
            std::exit(1);
        }
    }

    int64_t boardNegate(int64_t a)
    {
        const int64_t mask24 = (1ll << 24) - 1;
        return ((a & mask24) << 24) | ((a & (mask24 << 24)) >> 24);
    }
};

File: Malom3/PerfectPlayer.vb
' Malom, a Nine Men's Morris (and variants) player and solver program.
' Copyright(C) 2007-2016  Gabor E. Gevay, Gabor Danner
' 
' See our webpage (and the paper linked from there):
' http://compalg.inf.elte.hu/~ggevay/mills/index.php
' 
' 
' This program is free software: you can redistribute it and/or modify
' it under the terms of the GNU General Public License as published by
' the Free Software Foundation, either version 3 of the License, or
' (at your option) any later version.
' 
' This program is distributed in the hope that it will be useful,
' but WITHOUT ANY WARRANTY; without even the implied warranty of
' MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
' GNU General Public License for more details.
' 
' You should have received a copy of the GNU General Public License
' along with this program.  If not, see <http://www.gnu.org/licenses/>.


Imports System.IO
Imports Wrappers

Public Class Sectors
    Public Shared sectors As New Dictionary(Of id, Sector)
    Shared created As Boolean = False

    <System.Runtime.ExceptionServices.HandleProcessCorruptedStateExceptionsAttribute>
    Public Shared Function getsectors() As Dictionary(Of id, Sector)
        Try

            If Not created Then

                Wrappers.Init.init_sym_lookuptables()
                Wrappers.Init.init_sec_vals()

                For w = 0 To Rules.MaxKSZ
                    For b = 0 To Rules.MaxKSZ
                        For wf = 0 To Rules.MaxKSZ
                            For bf = 0 To Rules.MaxKSZ
                                Dim fname = String.Format(Rules.VariantName & "_{0}_{1}_{2}_{3}.sec" & Constants.Fname_suffix, w, b, wf, bf)
                                'Console.WriteLine("Looking for database file " & fname)
                                Dim id As New id(w, b, wf, bf)
                                If File.Exists(fname) Then
                                    sectors(id) = New Sector(id)
                                End If
                            Next
                        Next
                    Next
                Next

                created = True
            End If

            Return sectors

        Catch ex As Exception
            If TypeOf ex Is KeyNotFoundException Then Throw
            MsgBox("An error happened in getsectors" & vbCrLf & ex.ToString, MsgBoxStyle.Critical)
            Environment.Exit(1)
            Return Nothing
        End Try
    End Function

    Public Shared Function HasDatabase() As Boolean
        Return getsectors().Count > 0
    End Function
End Class

Public Class PerfectPlayer
    Inherits Player

    Dim secs As New Dictionary(Of id, Sector)
    Const UseWRGM = False 'Ha ez True, akkor az engine-ben korlatozni kell a melyseget (ez mostmar automatikus)
    Dim Eng As Engine

    Public Sub New()
        Debug.Assert(Sectors.HasDatabase)
        secs = Sectors.getsectors()

        If UseWRGM Then
            Debug.Assert(Rules.AlphaBetaAvailable)
            Eng = New Engine(Nothing, Nothing, Nothing, True)
            Eng.InitEngine()
        End If
    End Sub

    Public Overrides Sub Enter(ByVal _g As Game)
        MyBase.Enter(_g)
    End Sub

    Public Overrides Sub Quit()
        MyBase.Quit()
    End Sub

    <System.Runtime.ExceptionServices.HandleProcessCorruptedStateExceptionsAttribute>
    Public Function GetSec(s As GameState) As Sector
        Try
            If s.KLE Then Return Nothing

            Dim id As New id(s.StoneCount(0), s.StoneCount(1), Rules.MaxKSZ - s.SetStoneCount(0), Rules.MaxKSZ - s.SetStoneCount(1))

            If s.SideToMove = 1 Then
                id.negate()
            End If

            Return secs(id)
        Catch ex As Exception
            If TypeOf ex Is KeyNotFoundException Then Throw
            MsgBox("An error happened in GetSec" & vbCrLf & ex.ToString, MsgBoxStyle.Critical)
            Environment.Exit(1)
            Return Nothing
        End Try
    End Function

    Private Function NegateState(s As GameState) As GameState
        Dim r As New GameState(s)
        For i = 0 To 23
            If r.T(i) = 0 Then
                r.T(i) = 1
            ElseIf r.T(i) = 1 Then
                r.T(i) = 0
            End If
        Next
        r.StoneCount(0) = s.StoneCount(1)
        r.StoneCount(1) = s.StoneCount(0)
        r.SetStoneCount(0) = s.SetStoneCount(1)
        r.SetStoneCount(1) = s.SetStoneCount(0)
        Return r
    End Function

    <System.Runtime.ExceptionServices.HandleProcessCorruptedStateExceptionsAttribute>
    Public Shared Function ToHumanReadableEval(e As gui_eval_elem2) As String
        Try
            Return e.ToString()
        Catch ex As Exception
            MsgBox("An error happened in ToHumanReadableEval" & vbCrLf & ex.ToString, MsgBoxStyle.Critical)
            Environment.Exit(1)
            Return Nothing
        End Try
    End Function



    Public Enum MoveType
        SetMove
        SlideMove 'should be renamed to SlideOrJumpMove
    End Enum

    Structure Move
        Public hon, hov As Integer
        Public MoveType As MoveType
        Public WithTaking, OnlyTaking As Boolean 'A With-esben benne van a malombecsukas lepese is, az Only-sban pedig csak a levetel
        Public TakeHon As Integer

        Public Function ToBitBoard() As Integer
            If OnlyTaking Then
                Return 1 << TakeHon
            End If
            Dim ret As Integer = 1 << hov
            If MoveType = MoveType.SlideMove Then
                ret += 1 << hon
            End If
            If WithTaking Then
                ret += 1 << TakeHon
            End If
            Return ret
        End Function
    End Structure

    Private Function FutureKorongCount(s As GameState) As Integer
        Return s.StoneCount(s.SideToMove) + Rules.MaxKSZ - s.SetStoneCount(s.SideToMove) 'TODO: refactor to call to FutureStoneCount
    End Function

    Private Function MakesMill(ByVal s As GameState, ByVal hon As Integer, ByVal hov As Integer) As Boolean
        Dim s2 As New GameState(s)
        If hon <> -1 Then s2.T(hon) = -1
        s2.T(hov) = s.SideToMove
        Return -1 <> Rules.Malome(hov, s2)
    End Function

    Private Function IsMill(ByVal s As GameState, ByVal m As Integer) As Boolean
        Return -1 <> Rules.Malome(m, s)
    End Function

    Private Function SetMoves(ByVal s As GameState) As List(Of Move)
        Dim r As New List(Of Move)
        For i = 0 To 23
            If s.T(i) = -1 Then
                r.Add(New Move With {.hov = i, .WithTaking = MakesMill(s, -1, i), .MoveType = MoveType.SetMove})
            End If
        Next
        Return r
    End Function
    Private Function SlideMoves(ByVal s As GameState) As List(Of Move)
        Dim r As New List(Of Move)
        For i = 0 To 23
            For j = 0 To 23
                If s.T(i) = s.SideToMove And s.T(j) = -1 And (FutureKorongCount(s) = 3 Or Rules.BoardGraph(i, j)) Then
                    r.Add(New Move With {.hon = i, .hov = j, .WithTaking = MakesMill(s, i, j), .MoveType = MoveType.SlideMove})
                End If
            Next
        Next
        Return r
    End Function

    'm-ben egy WithTakinges lepes van, aminek nincsen kitoltve a TakeHonja. Ez a fuggveny keszit egy listat, aminek az elemei m masolatai, kiegeszitve egy-egy lehetseges levetellel
    Private Function WithTakingMoves(ByVal s As GameState, ByVal m As Move) As List(Of Move)
        Dim r As New List(Of Move)

        Dim EverythingInMill As Boolean = True
        For i = 0 To 23
            If s.T(i) = 1 - s.SideToMove And Not IsMill(s, i) Then EverythingInMill = False
        Next

        For i = 0 To 23
            If s.T(i) = 1 - s.SideToMove And (Not IsMill(s, i) Or EverythingInMill) Then
                Dim m2 = m
                m2.TakeHon = i
                r.Add(m2)
            End If
        Next

        Return r
    End Function
    Private Function OnlyTakingMoves(ByVal s As GameState) As List(Of Move) 'itt van nemi copy-paste kod
        Dim r As New List(Of Move)

        Dim EverythingInMill As Boolean = True
        For i = 0 To 23
            If s.T(i) = 1 - s.SideToMove And Not IsMill(s, i) Then EverythingInMill = False
        Next

        For i = 0 To 23
            If s.T(i) = 1 - s.SideToMove And (Not IsMill(s, i) Or EverythingInMill) Then
                r.Add(New Move With {.OnlyTaking = True, .TakeHon = i})
            End If
        Next

        Return r
    End Function


    Public Function GetMoveList(ByVal s As GameState) As List(Of Move)
        Dim ms0, ms As List(Of Move)
        If Not s.KLE Then
            If Wrappers.Constants.Variant = Wrappers.Constants.Variants.std Or Wrappers.Constants.Variant = Wrappers.Constants.Variants.mora Then
                If s.SetStoneCount(s.SideToMove) < Rules.MaxKSZ Then
                    ms0 = SetMoves(s)
                Else
                    ms0 = SlideMoves(s)
                End If
            Else 'Lasker
                ms0 = SlideMoves(s)
                If s.SetStoneCount(s.SideToMove) < Rules.MaxKSZ Then
                    ms0.AddRange(SetMoves(s))
                End If
            End If

            ms = New List(Of Move)
            For i = 0 To ms0.Count - 1
                If Not ms0(i).WithTaking Then
                    ms.Add(ms0(i))
                Else
                    ms.AddRange(WithTakingMoves(s, ms0(i)))
                End If
            Next
        Else 'KLE
            ms = OnlyTakingMoves(s)
        End If

        Return ms
    End Function

    Private Function MakeMoveInState(ByVal s As GameState, ByVal m As Move) As GameState
        Dim s2 As New GameState(s)
        If Not m.OnlyTaking Then
            If m.MoveType = MoveType.SetMove Then
                s2.MakeMove(New SetKorong(m.hov))
            Else
                s2.MakeMove(New MoveKorong(m.hon, m.hov))
            End If
            If m.WithTaking Then s2.MakeMove(New LeveszKorong(m.TakeHon))
        Else
            s2.MakeMove(New LeveszKorong(m.TakeHon))
        End If
        Return s2
    End Function

    <System.Runtime.ExceptionServices.HandleProcessCorruptedStateExceptionsAttribute>
    Public Function MoveValue(ByVal s As GameState, ByVal m As Move) As gui_eval_elem2
        Try
            Return Eval(MakeMoveInState(s, m)).undo_negate(GetSec(s))
        Catch ex As Exception
            If TypeOf ex Is KeyNotFoundException Then Throw
            MsgBox("Exception in MoveValue" & vbCrLf & ex.ToString, MsgBoxStyle.Critical)
            Environment.Exit(1)
        End Try
    End Function

    'Returns all elements of l, that are maximal by f
    Public Shared Function AllMaxBy(Of T, K As IComparable(Of K))(ByVal f As Func(Of T, K), ByVal l As List(Of T), MinValue As K) As List(Of T)
        Dim r As New List(Of T)
        'Dim ma = Integer.MinValue
        'Dim ma As K = f(l(0))
        Dim ma = MinValue
        For Each m In l
            Dim e = f(m)
            'If e > ma Then
            If e.CompareTo(ma) > 0 Then
                ma = e
                r.Clear()
                r.Add(m)
            Else
                If e.CompareTo(ma) = 0 Then
                    r.Add(m)
                End If
            End If
        Next
        Return r
    End Function

    Public Function GoodMoves(ByVal s As GameState) As List(Of Move)
        Return AllMaxBy(Function(m) MoveValue(s, m), GetMoveList(s), gui_eval_elem2.min_value(GetSec(s)))
    End Function

    Private Function NGMAfterMove(ByVal s As GameState, ByVal m As Move) As Integer
        Return NumGoodMoves(MakeMoveInState(s, m))
    End Function

    Private Function MinNGMMoves(ByVal s As GameState) As List(Of Move)
        Return AllMaxBy(Function(m) -NGMAfterMove(s, m), GoodMoves(s), Integer.MinValue)
    End Function

    Private Shared rnd As New Random()
    Public Function ChooseRandom(Of T)(ByVal l As List(Of T)) As T
        Return l(rnd.Next(l.Count))
    End Function

    Private Sub SendMoveToGUI(m As Move)
        If Not m.OnlyTaking Then
            If m.MoveType = MoveType.SetMove Then
                G.MakeMove(New SetKorong(m.hov))
            Else
                G.MakeMove(New MoveKorong(m.hon, m.hov))
            End If
        Else
            G.MakeMove(New LeveszKorong(m.TakeHon))
        End If
    End Sub

    'A koronglevetel kezelese:
    '-Az eval nem tud KLE allast kezelni, mivel a solver szempontjabol ezek nem leteznek.
    '-Tehat az itteni move strukturaban ossze van vonva a malombecsukas a koronglevetelevel.
    '-Viszont a GUI szempontjabol ez kulon esemeny.
    '-Vegulis az lett, hogy amikor a GUI-tol KLE allast kapunk, akkor ujra belenezunk az adatbazisba.
    '   (Igy tudjuk kezelni, ha pl. bepaste-elnek KLE allast (valamint jobban illeszkedik a regi engine-hez))
    <System.Runtime.ExceptionServices.HandleProcessCorruptedStateExceptionsAttribute>
    Public Overrides Sub ToMove(ByVal s As GameState)
        Try
            'Dim mh = GoodMoves(s)(0)
            'Dim mh = ChooseRandom(GoodMoves(s))
            'Dim mh = MinNGMMoves(s)(0)
            'Dim mh = RecWRGM(s).m
            'Dim mh = If(UseWRGM, ChooseRandom(MinWRGMMoves(s)), ChooseRandom(MinNGMMoves(s)))
            'Dim mh = MinNGMMoves(s)(0)

            'Dim mh As Move
            'If s.LépésCount > 6 Then
            '    mh = MinNGMMoves(s)(0)
            'Else
            '    mh = ChooseNonNTREKSMoveIfPossible(s)
            'End If

            'Dim mh = ChooseNonNTREKS8998MoveIfPossible(s)


            Dim mh = ChooseRandom(GoodMoves(s))
            'Dim mh = ChooseRandom(MinNGMMoves(s))


            'Dim ma = MoveValue(s, mh)

            'Main.LblPerfEval.Text = "Ev: " & ma

            SendMoveToGUI(mh)
        Catch ex As KeyNotFoundException
            'Debug.Assert(Not s.KLE) 'kivettuk, de nem tudjuk, hogy miert volt bent
            SendMoveToGUI(ChooseRandom(GetMoveList(s)))
        Catch ex As Exception
            MsgBox("Exception in ToMove" & vbCrLf & ex.ToString, MsgBoxStyle.Critical)
            Environment.Exit(1)
        End Try
    End Sub


    Private Function NumGoodMoves(ByVal s As GameState) As Integer
        If FutureKorongCount(s) < 3 Then Return 0
        Dim ma = gui_eval_elem2.min_value(GetSec(s)), mh As Move, c = 0
        For Each m In GetMoveList(s)
            Dim e = MoveValue(s, m)
            If e > ma Then
                ma = e
                mh = m
                c = 1
            Else
                If e = ma Then c += 1
            End If
        Next
        Return c
    End Function

    Dim cp As Integer
    Structure MoveValuePair
        Dim m As Move
        Dim val As Double
    End Structure
    Const WRGMInf As Double = 2 'jo ez?

    Private Shared EvalLock As New Object
    <System.Runtime.ExceptionServices.HandleProcessCorruptedStateExceptionsAttribute>
    Private Function Eval(ByVal s As GameState) As gui_eval_elem2
        Try
            SyncLock EvalLock 'ez azert kell, mert ha mindket jatekos hivogatja (az egyik esetleg OppTime), akkor a hash objektumokkal gond lehetne a dinamikus cserelgetes miatt
                Debug.Assert(Not s.KLE)
                Dim id As New id(s.StoneCount(0), s.StoneCount(1), Rules.MaxKSZ - s.SetStoneCount(0), Rules.MaxKSZ - s.SetStoneCount(1))

                If FutureKorongCount(s) < 3 Then Return gui_eval_elem2.virt_loss_val

                Dim a As Int64
                For i = 0 To 23
                    If s.T(i) = 0 Then
                        a = a Or (CType(1, Int64) << i)
                    ElseIf s.T(i) = 1 Then
                        a = a Or (CType(1, Int64) << (i + 24))
                    End If
                Next

                If s.SideToMove = 1 Then
                    a = BoardNegate(a)
                    id.negate()
                End If

                Dim sec = secs(id)

                Return sec.hash(a).Item2
            End SyncLock
        Catch ex As Exception
            If TypeOf ex Is KeyNotFoundException Then Throw
            MsgBox("Exception in Eval" & vbCrLf & ex.ToString, MsgBoxStyle.Critical)
            Environment.Exit(1)
        End Try
    End Function


    Private Function BoardNegate(ByVal a As Int64) As Int64
        Const mask24 As Int64 = (1 << 24) - 1
        Return ((a And mask24) << 24) Or ((a And (mask24 << 24)) >> 24)
    End Function
End Class
